src/backend/distributed/commands/alter_table.c

```diff
@@ -1589,7 +1625,36 @@ ReplaceTable(Oid sourceId, Oid targetId, List *justBeforeDropCommands,
 		ExecuteQueryViaSPI(query->data, SPI_OK_INSERT);
 	}
 
-	List *ownedSequences = getOwnedSequences(sourceId);
+	/* We need to drop sequence dependencies for identities to preserve their
+	 * last value on the workers if it's a undistribute table conversion
+	 */
+	if (conversionType == UNDISTRIBUTE_TABLE)
+	{
+		List *ownedIdentitySequences = getOwnedSequences_internal(sourceId, 0,
+																  DEPENDENCY_INTERNAL);
+		if (ownedIdentitySequences != NIL && ShouldSyncTableMetadata(sourceId))
+		{
+			char *qualifiedTableName = quote_qualified_identifier(schemaName, sourceName);
+
+			/*
+			 * We are converting a citus local table to a distributed/reference table,
+			 * so we should prevent dropping the sequence on the table. Otherwise, we'd
+			 * lose track of the previous changes in the sequence.
+			 */
+			StringInfo command = makeStringInfo();
+
+			appendStringInfo(command,
+							 "SELECT pg_catalog.worker_drop_sequence_dependency(%s);",
```

Comment:
The reason why this provides "kind of" a solution to the issue @marcocitus pointed out with `undistribute_table`, is that we use `CREATE SEQUENCE IF NOT EXISTS` statements when we are distributing the table again. However, this hacky solution is broken if we simply rename the sequence before re-distributing, and then we still get the same issue:
```SQL
CREATE TABLE color (
    color_id BIGINT GENERATED ALWAYS AS IDENTITY UNIQUE,
    color_name VARCHAR NOT NULL
);
select create_distributed_table('color', 'color_id');

-- from worker node:
INSERT INTO color(color_name) VALUES ('Red');
NOTICE:  executing the command locally: INSERT INTO public.color_102349 (color_id, color_name) VALUES ('281474976710657'::bigint, 'Red'::character varying)     
                  
-- back on coordinator
select undistribute_table('color');

ALTER SEQUENCE color_color_id_seq RENAME TO myseq;

select create_distributed_table('color', 'color_id');

-- from worker node:
INSERT INTO color(color_name) VALUES ('Red');
ERROR:  duplicate key value violates unique constraint "color_color_id_key_102252"
DETAIL:  Key (color_id)=(281474976710657) already exists.
```

Reply:
Note that we don't have this issue with serial pseudotypes, because the serial sequence is considered distributed even after we undistribute the table. Hence, the renaming propagates.

Reply:
Note from Teams chat: it is because for serial sequences we are re-using them when recreating the table, because we don't recreate the table with `serial` anymore: we recreate the table with `default nextval()` which allows to specify the old sequence. But for generated identities we are creating a new sequence which is not marked as distributed.

Reply:
I feel like it might be better not to attempt to solve this right now (as in, to remove this code), because there are probably a whole bunch of things that could go wrong here (e.g. what if the column is renamed before we redistribute?). It seems slightly preferable to keep our database in a coherent state and live with the usability issue until we find a better solution.

Reply:
marking the sequence as distribute is probably a good thing to do, though not sure it solves all of our undistribute_table problems.

Reply:
Hmmm, actually we would still have a problem if the user had already renamed the identity sequence before undistributing the table.

Reply:
> marking it as distributed right away at create_distributed_table time would help there

That's already the case.

But the problem currently is that at `RenameIdentitiesOnTable`, Postgres generated names are used.

Reply:
well, we should definitely not use ChooseRelationName in RenameIdentitiesOnTable